pkg/coremodel/playlist/coremodel.cue

```diff
@@ -0,0 +1,161 @@
+				interval: string | *"5m"
+				// The ordered list of items that the playlist will iterate over.
+				items?: [...#Item]
+
+				///////////////////////////////////////
+				// Definitions (referenced above) are declared below
+
+				#Type: "dashboard_by_id" | "dashboard_by_tag" | "dashboard_by_uid" @cuetsy(kind="enum",memberNames="DashboardById|DashboardByTag|DashboardByUID")
+
+				#Item: {
+					// Type of the item.
+					type: #Type
+					// Value depends on type and describes the playlist item.
+					//
+					//  - dashboard_by_id: The value is an internal numerical identifier set by Grafana. This
+					//  is not portable as the numerical identifier is non-deterministic between different instances. Deprecated.
+					//  - dashboard_by_tag: The value is a tag which is set on any number of dashboards. All
+					//  dashboards behind the tag will be added to the playlist.
+					//  - dashboard_by_uid: The value is the UID identifier of a dashboard.
+					value: string
+					// Human-friendly title for item's listing in playlist.
+					//
+					// Deprecated.
+					title: string
+				} @cuetsy(kind="interface")
+			},
+		]
+	},
+	{
+		schemas: [
+			{//1.0 - start of new types
+				// Unique playlist identifier. Generated on creation, either by the
+				// creator of the playlist of by the application.
+				uid: string
+				// Name of the playlist.
+				name: string
+				// Interval sets the time between switching views in a playlist.
+				// FIXME: Is this based on a standardized format or what options are available? Can datemath be used?
+				interval: string | *"5m"
+				// The ordered list of items that the playlist will iterate over.
+				items?: [...#Item]
+
+				///////////////////////////////////////
+				// Definitions (referenced above) are declared below
+
+				#Type: "dashboard_by_tag" | "dashboard_by_uid" @cuetsy(kind="enum",memberNames="DashboardByTag|DashboardByUID")
+
+				#Item: {
+					// Type of the item.
+					type: #Type
+					// Value depends on type and describes the playlist item.
+					//
+					//  - dashboard_by_tag: The value is a tag which is set on any number of dashboards. All
+					//  dashboards behind the tag will be added to the playlist.
+					//  - dashboard_by_uid: The value is the UID identifier of a dashboard.
+					value: string
+				}
+			},
+		]
+
+		lens: forward: {
+			to:         seqs[1].schemas[0]
+			from:       seqs[0].schemas[0]
+			translated: to & rel
+			rel: {
+				uid:      from.uid
+				name:     from.name
+				interval: from.interval
+				if (from.items != _|_) {
+					items: [ for item in from.items {
+						if (item.type == "dashboard_by_tag" || item.type == "dashboard_by_uid") {
+							type: item.type
+							value: item.value
+						}
+						if (item.type == "dashboard_by_id") {
+							type:  "dashboard_by_uid"
+							value: PlaceholderPrefix + item.value
+						}
+					}]
+				}
+			}
+			lacunas: [
```

Comment:
I don't get how any of this is helpful -- I get that it allows the old and new models, but also is exactly equivalent to just ignoring any of the now unused values.

Reply:
Note also the frontend will migrate any `dashboard_by_id` to `dashboard_by_uid` using real values from the databases:
https://github.com/grafana/grafana/blob/main/public/app/features/playlist/api.ts#L32

Reply:
Correct, this is the exact case where we can't do anything reasonable, because there's no database access here.

The frontend can still migrate. While we do use a lacuna to indicate the existence of this placeholder value, and that is how we'd generally expect backend code to handle it, we _also_ structure the value in the translated itself that the frontend can sniff for it.

That is, if there's an input item like this:

```json
{
  "type": "dashboard_by_id",
  "value": "42"
  "title": "something"
}
```

then, as currently written, the translated output will be:

```json
{
  "type": "dashboard_by_id",
  "value": "placeholder-cupcake-42"
  "title": "something"
}
```

This approach needs an actual convention - we were just playing around with `cupcake` - but it's a pattern we can reuse anywhere we have to bridge between an id-driven and uid-driven world.

The advantage is that same boring, quiet goal for all of this: a clear, consistent shared contract about what the values in the objects are supposed to be, and making that mechanism available to at least any Go program. A terraform provider can warn the user that this is the lossy translation that _will_ happen, while accepting either version of the schema. If provisioning a new Grafana instance, it could choose to reject any such input altogether because it can know _a priori_ that there's no way that the `42` decodable from `placeholder-cupcake-42` will actually refer to the originally intended dashboard. That could easily save hours of work and confusion, if the user is even aware _at all_ that the reference doesn't line up.

Reply:
i realize i buried the lede - our premise is that putting a placeholder value in _is reasonable_, because the backend can then sniff the value for the placeholder pattern, query the database, and replace it on the fly. (Maybe save the result, too? 🤷🏻 )

Either way, this _should_ mean that the frontend never has to consider doing such a migration.

Reply:
That is probably where we disagree the most -- needing to run a migration and *then* keep checking if you still need to actually migrate is pretty terrible.  I think that is why most people end up writing migrations programmatically.  I have been punting thinking we could think about this later, but first one out of the gate!